### PR TITLE
[expo-cli] Migrate build command to TypeScript

### PR DESCRIPTION
**Pull request by EvanBacon**

### CODE REVIEW

File: packages/expo-cli/src/commands/build/BaseBuilder.ts [hunk 1]
```diff
@@ -78,7 +57,7 @@ export default class BaseBuilder {
   }
 
   async prepareProjectInfo(): Promise<void> {
-    // always use local json to unify behaviour between regular apps and self hosted ones
+    // always use local json to unify behavior between regular apps and self hosted ones
```
Comment (brentvatne) [inline packages/expo-cli/src/commands/build/BaseBuilder.ts]: it's behaviour everywhere except USA!

Comment (EvanBacon) [reply packages/expo-cli/src/commands/build/BaseBuilder.ts]: I have a spellcheck linter installed in code. [also](https://i.imgflip.com/3ggvsj.png)

File: packages/expo-cli/src/commands/build/BaseBuilder.ts [hunk 2]
```diff
@@ -125,17 +104,17 @@ export default class BaseBuilder {
         releaseChannel: this.options.releaseChannel,
         publicUrl: this.options.publicUrl,
         sdkVersion: this.manifest.sdkVersion,
-      });
+      } as any);
     }
-    if (buildStatus.jobs && buildStatus.jobs.length) {
+    if ('jobs' in buildStatus && Array.isArray(buildStatus.jobs) && buildStatus.jobs.length) {
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build/BaseBuilder.ts]:
```suggestion
    if (buildStatus?.jobs?.length > 0) {
```

File: packages/expo-cli/src/commands/build/BaseBuilder.ts [hunk 3]
```diff
@@ -150,26 +129,33 @@ export default class BaseBuilder {
         releaseChannel: this.options.releaseChannel,
       });
     }
-    if (buildStatus.err) {
+    if ('err' in buildStatus && buildStatus.err) {
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build/BaseBuilder.ts]:
```suggestion
    if (buildStatus?.err) {
```

File: packages/expo-cli/src/commands/build/ios/utils/image.ts [hunk 4]
```diff
@@ -19,7 +19,7 @@ async function ensurePNGIsNotTransparent(imagePathOrURL) {
       .on('metadata', ({ alpha }) => {
         if (!alpha) {
           hasAlreadyResolved = true;
-          stream.close();
+          if ('close' in stream) stream.close();
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build/ios/utils/image.ts]:
```suggestion
         stream?.close();
```

File: packages/expo-cli/src/commands/build/ios/credentials/revoke.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+): Promise<void> {
+  if (manager instanceof ProvisioningProfileManager) {
+    const { bundleIdentifier } = projectMetadata;
+    await manager.revoke(bundleIdentifier!);
+    return;
+  }
+
+  if (manager.list) {
+    const list = await manager.list();
+    if (!list || list.length === 0) {
```
Comment (wkozyra95) [inline packages/expo-cli/src/commands/build/ios/credentials/revoke.ts]:
```suggestion
    if (!list?.length) {
```

File: packages/expo-cli/src/commands/build/BaseBuilder.ts (hunk 3 above)
Comment (EvanBacon) [reply packages/expo-cli/src/commands/build/BaseBuilder.ts]:
```
Property 'err' does not exist on type 'BuildStatusResult | BuildCreatedResult'.
  Property 'err' does not exist on type 'BuildCreatedResult'.
```

File: packages/expo-cli/src/commands/build/ios/utils/image.ts (hunk 4 above)
Comment (EvanBacon) [reply packages/expo-cli/src/commands/build/ios/utils/image.ts]:
```
Property 'close' does not exist on type 'Request | ReadStream'.
  Property 'close' does not exist on type 'Request'.
```

File: packages/expo-cli/src/commands/build/ios/IOSBuilder.ts [hunk 6]
```diff
@@ -50,7 +53,7 @@ See https://docs.expo.io/versions/latest/distribution/building-standalone-apps/#
   }
 
   async prepareCredentials() {
-    const username = this.manifest.owner || this.user.username;
+    const username = this.manifest.owner || this.user?.username!;
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-cli/src/commands/build/ios/IOSBuilder.ts]:
```
Type 'undefined' is not assignable to type 'string'.
```

**Pull request merged**